libs/platform/core/src/services/query/core/core-query.ts

```diff
@@ -0,0 +1,233 @@
+          subject$.next(initialState);
+        }
+      })
+    );
+
+    const refreshTrigger$ = this.getTriggerStream(this.options.refreshOn).pipe(
+      tap(() => {
+        if (!subject$.value.stale) {
+          subject$.next({
+            error: subject$.value.error,
+            loading: subject$.value.loading,
+            stale: true,
+            data: subject$.value.data,
+          });
+        }
+      })
+    );
+
+    const loadLogic$ = needToLoad$.pipe(
+      tap(() => {
+        if (subject$.value.loading!) {
+          subject$.next({
+            error: subject$.value.error,
+            loading: true,
+            stale: subject$.value.stale,
+            data: subject$.value.data,
+          });
+        }
+      }),
+      switchMap(() =>
+        from(this.options.loader(qualifier)).pipe(takeUntil(resetTrigger$))
+      ),
+      tap((data) => {
+        this.onLoad(data, qualifier);
+        subject$.next({ error: false, loading: false, stale: false, data });
+      }),
+      catchError((error, source$) => {
+        this.onError(error, qualifier);
+        subject$.next({ error, loading: false, stale: false, data: undefined });
+        return source$;
+      })
+    );
+
+    let offlineSub: Subscription | undefined;
+
+    const state$ = using(
+      () => {
+        offlineSub?.unsubscribe();
+
+        const subscription = loadLogic$.subscribe();
+        subscription.add(() => resetTrigger$.subscribe());
+
+        return {
+          unsubscribe: () => {
+            subscription.unsubscribe();
+
+            if (this.options.volatile) {
+              // if query is volatile set it to initial state
+              subject$.next(initialState);
+              return;
+            }
+
+            offlineSub = merge(refreshTrigger$, resetTrigger$)
+              .pipe(take(1), takeUntil(this.destroyNotifier$ ?? NEVER))
+              .subscribe(() => subject$.next(initialState));
+          },
+        };
+      },
+      () => subject$
+    ).pipe(shareReplay({ bufferSize: 1, refCount: !this.options.permanent }));
+
+    const data$ = state$.pipe(
+      map((x) => x.data),
+      distinctUntilChanged(),
+      shareReplay({ bufferSize: 1, refCount: true })
+    );
+
+    this.queryCache.set(key, { subject$, state$, data$ });
+  }
+
+  get(qualifier: Qualifier): Observable<ValueType | undefined> {
+    return this.queryCache.get(this.prepareKey(qualifier))!.data$;
+  }
+
+  getState(qualifier: Qualifier): Observable<QueryState<ValueType>> {
+    return this.queryCache.get(this.prepareKey(qualifier))!.state$;
+  }
+
+  set({
+    data,
+    qualifier,
+    optimistic,
+  }: {
+    data: ValueType;
+    qualifier: Qualifier;
+    optimistic?: boolean;
+  }): void {
+    const subject$ = this.queryCache.get(this.prepareKey(qualifier))!.subject$;
+    subject$.next({
+      error: subject$.value.error,
+      loading: subject$.value.loading,
+      stale: optimistic ? true : subject$.value.stale,
```

Comment:
```suggestion
      stale: optimistic || subject$.value.stale,
```

Reply:
Hmm... 
I like the first one for readibility, and hope that optimizer will optimize it. But not strong opinion :)